File: app/src/main/java/com/example/a2ui/chat/data/repository/RealChatRepository.kt
package com.example.a2ui.chat.data.repository

import android.util.Log
import com.example.a2ui.chat.data.model.AgentResponseDto
import com.example.a2ui.chat.data.model.toDomain
import com.example.a2ui.chat.domain.model.Message
import com.example.a2ui.chat.domain.model.Sender
import com.example.a2ui.chat.domain.repository.ChatRepository
import com.example.a2ui.chat.domain.repository.StreamEvent
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.withContext
import java.io.IOException
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.Calendar
import java.util.UUID
import java.util.concurrent.TimeUnit

@Serializable
private data class ChatRequest(val message: String)

@Serializable
private data class TokenData(val token: String)

class RealChatRepository(
    private val baseUrl: String = "http://10.0.2.2:8000"
) : ChatRepository {

    companion object {
        private const val TAG = "A2UI.Repo"
    }

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    /** Longer read timeout for streaming — tokens may arrive slowly. */
    private val streamingClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    override suspend fun sendMessage(userMessage: String): Message {
        Log.i(TAG, "sendMessage: \"${userMessage.take(60)}\"")
        return try {
            val requestBody = json.encodeToString(ChatRequest(message = userMessage))
                .toRequestBody(jsonMediaType)

            val request = Request.Builder()
                .url("$baseUrl/chat")
                .post(requestBody)
                .build()

            val responseBody = withContext(Dispatchers.IO) {
                client.newCall(request).execute().use { response ->
                    Log.d(TAG, "sendMessage: HTTP ${response.code}")
                    response.body?.string()
                        ?: throw IllegalStateException("Empty response body")
                }
            }

            val agentResponse = json.decodeFromString<AgentResponseDto>(responseBody)
            Log.d(TAG, "sendMessage: response hasUi=${agentResponse.uiDefinition != null}")

            Message(
                id = UUID.randomUUID().toString(),
                content = agentResponse.text,
                sender = Sender.AI,
                timestamp = System.currentTimeMillis(),
                isLoading = false,
                uiDefinition = agentResponse.uiDefinition?.toDomain()
            )
        } catch (e: Exception) {
            Log.e(TAG, "sendMessage: failed", e)
            Message(
                id = UUID.randomUUID().toString(),
                content = "Couldn't reach the agent server. Make sure it's running at $baseUrl.",
                sender = Sender.AI,
                timestamp = System.currentTimeMillis(),
                isLoading = false
            )
        }
    }

    override fun sendMessageStream(userMessage: String): Flow<StreamEvent> = flow {
        Log.i(TAG, "[stream] start: \"${userMessage.take(60)}\"")

        var attempt = 0
        val maxAttempts = 3
        var lastError: IOException? = null
        var doneReceived = false

        while (attempt < maxAttempts && !doneReceived) {
            if (attempt > 0) {
                val delayMs = (2_000L shl (attempt - 1)) // 2s, 4s, 8s
                Log.w(TAG, "[stream] retry attempt=$attempt after ${delayMs}ms")
                delay(delayMs)
            }
            attempt++
            try {
                val requestBody = json.encodeToString(ChatRequest(message = userMessage))
                    .toRequestBody(jsonMediaType)

                val request = Request.Builder()
                    .url("$baseUrl/chat/stream")
                    .post(requestBody)
                    .header("Accept", "text/event-stream")
                    .build()

                val response = withContext(Dispatchers.IO) {
                    streamingClient.newCall(request).execute()
                }

                Log.d(TAG, "[stream] HTTP ${response.code}")

                if (!response.isSuccessful) {
                    response.close()
                    throw IllegalStateException("HTTP ${response.code}: ${response.message}")
                }

                val source = response.body?.source()
                if (source == null) {
                    response.close()
                    throw IllegalStateException("Empty response body")
                }

                try {
                    var eventType = ""
                    while (true) {
                        val line = withContext(Dispatchers.IO) {
                            if (source.exhausted()) null else source.readUtf8Line()
                        } ?: break

                        when {
                            line.startsWith("event: ") -> {
                                eventType = line.removePrefix("event: ").trim()
                            }
                            line.startsWith("data: ") -> {
                                val data = line.removePrefix("data: ")
                                Log.d(TAG, "[stream] event=$eventType dataLen=${data.length}")
                                when (eventType) {
                                    // ── A2UI v0.8 protocol events ──────────────
                                    "a2ui_op" -> {
                                        emit(StreamEvent.A2UiOp(data))
                                    }

                                    "text" -> {
                                        try {
                                            val textObj = json.parseToJsonElement(data).jsonObject
                                            val text = textObj["text"]?.jsonPrimitive?.contentOrNull ?: ""
                                            emit(StreamEvent.TextContent(text))
                                        } catch (e: Exception) {
                                            Log.w(TAG, "[stream] malformed text event: \"${data.take(200)}\"", e)
                                        }
                                    }

                                    "done" -> {
                                        Log.i(TAG, "[stream] done event received")
                                        // Check for backward-compatible "done" with embedded ui_definition
                                        val doneMessage = parseDoneEvent(data)
                                        emit(StreamEvent.Done(doneMessage))
                                        doneReceived = true
                                    }

                                    // ── Legacy token streaming (backward compat) ──
                                    "token" -> {
                                        try {
                                            val tokenData = json.decodeFromString<TokenData>(data)
                                            emit(StreamEvent.Token(tokenData.token))
                                        } catch (e: Exception) {
                                            Log.w(TAG, "[stream] malformed token event: \"${data.take(200)}\"", e)
                                        }
                                    }
                                }
                                eventType = ""
                            }
                        }
                        if (doneReceived) break
                    }
                } finally {
                    try { response.close() } catch (_: Exception) { /* already handled */ }
                }
            } catch (e: IOException) {
                lastError = e
                Log.w(TAG, "[stream] IOException on attempt $attempt", e)
                // loop continues for retry
            } catch (e: Exception) {
                // Non-retryable error (e.g. HTTP 4xx/5xx IllegalStateException)
                Log.e(TAG, "[stream] non-retryable error", e)
                emit(StreamEvent.Error(e.message ?: "Unknown streaming error"))
                return@flow
            }
        }

        if (!doneReceived) {
            emit(StreamEvent.Error("Connection failed after $maxAttempts attempts: ${lastError?.message}"))
        }
    }

    /**
     * Spec-compliant JSONL streaming via [/chat/stream/jsonl].
     *
     * All SSE events arrive as plain `data:` lines (no custom `event:` type).
     * Each line is a JSONL object; we dispatch on the top-level key:
     *   "text"            → StreamEvent.TextContent
     *   "surfaceUpdate"   → StreamEvent.A2UiOp (raw JSON forwarded to SurfaceStateManager)
     *   "dataModelUpdate" → StreamEvent.A2UiOp
     *   "beginRendering"  → StreamEvent.A2UiOp  (arrives last — triggers render)
     *   "done"            → StreamEvent.Done
     */
    override fun sendMessageStreamJsonl(userMessage: String): Flow<StreamEvent> = flow {
        Log.i(TAG, "[jsonl-stream] start: \"${userMessage.take(60)}\"")

        val requestBody = json.encodeToString(ChatRequest(message = userMessage))
            .toRequestBody(jsonMediaType)

        val request = Request.Builder()
            .url("$baseUrl/chat/stream/jsonl")
            .post(requestBody)
            .header("Accept", "text/event-stream")
            .header("Cache-Control", "no-cache")
            .build()

        try {
            val response = withContext(Dispatchers.IO) {
                streamingClient.newCall(request).execute()
            }
            Log.d(TAG, "[jsonl-stream] HTTP ${response.code}")
            if (!response.isSuccessful) {
                response.close()
                emit(StreamEvent.Error("HTTP ${response.code}: ${response.message}"))
                return@flow
            }
            val source = response.body?.source() ?: run {
                response.close()
                emit(StreamEvent.Error("Empty response body"))
                return@flow
            }
            try {
                var doneReceived = false
                while (!doneReceived) {
                    val line = withContext(Dispatchers.IO) {
                        if (source.exhausted()) null else source.readUtf8Line()
                    } ?: break

                    if (!line.startsWith("data: ")) continue
                    val data = line.removePrefix("data: ").trim()
                    if (data.isEmpty() || data == "{}") continue

                    try {
                        val obj = json.parseToJsonElement(data).jsonObject
                        when {
                            "text" in obj -> {
                                val text = obj["text"]?.jsonPrimitive?.contentOrNull ?: ""
                                Log.d(TAG, "[jsonl-stream] text: \"${text.take(60)}\"")
                                emit(StreamEvent.TextContent(text))
                            }
                            "surfaceUpdate" in obj || "dataModelUpdate" in obj || "beginRendering" in obj -> {
                                Log.d(TAG, "[jsonl-stream] a2ui op key=${obj.keys.firstOrNull()} dataLen=${data.length}")
                                emit(StreamEvent.A2UiOp(data))
                            }
                            "done" in obj -> {
                                Log.i(TAG, "[jsonl-stream] done received")
                                emit(StreamEvent.Done(Message(
                                    id = UUID.randomUUID().toString(),
                                    content = "",
                                    sender = Sender.AI,
                                    timestamp = System.currentTimeMillis(),
                                    isLoading = false
                                )))
                                doneReceived = true
                            }
                            else -> {
                                Log.w(TAG, "[jsonl-stream] unknown JSONL key: ${obj.keys}")
                            }
                        }
                    } catch (e: Exception) {
                        Log.w(TAG, "[jsonl-stream] parse error on line: \"${data.take(200)}\"", e)
                    }
                }
            } finally {
                try { response.close() } catch (_: Exception) {}
            }
        } catch (e: IOException) {
            Log.e(TAG, "[jsonl-stream] IO error", e)
            emit(StreamEvent.Error(e.message ?: "IO error"))
        } catch (e: Exception) {
            Log.e(TAG, "[jsonl-stream] error", e)
            emit(StreamEvent.Error(e.message ?: "Unknown error"))
        }
    }

    /**
     * Parse a `done` event payload. Handles two formats:
     * 1. **Legacy (snapshot mode):** data contains `text` and optional `ui_definition`.
     * 2. **A2UI v0.8:** data is `{}` — the ViewModel builds the final message from
     *    accumulated operations, so we return a shell message here.
     */
    private fun parseDoneEvent(data: String): Message {
        return try {
            val parsed = json.parseToJsonElement(data).jsonObject
            val text = parsed["text"]?.jsonPrimitive?.contentOrNull ?: ""
            val uiDefElement = parsed["ui_definition"]

            // Legacy format: done event carries an embedded ui_definition
            val uiDefinitionDto = if (uiDefElement != null && uiDefElement !is JsonNull) {
                json.decodeFromString<com.example.a2ui.chat.data.model.UiDefinitionDto>(
                    uiDefElement.toString()
                )
            } else null

            Message(
                id = UUID.randomUUID().toString(),
                content = text,
                sender = Sender.AI,
                timestamp = System.currentTimeMillis(),
                isLoading = false,
                uiDefinition = uiDefinitionDto?.toDomain()
            )
        } catch (e: Exception) {
            Message(
                id = UUID.randomUUID().toString(),
                content = data,
                sender = Sender.AI,
                timestamp = System.currentTimeMillis(),
                isLoading = false
            )
        }
    }

    // ── Event sending (Client → Server) ────────────────────────────────

    override suspend fun sendEvent(
        surfaceId: String,
        eventType: String,
        name: String?,
        sourceComponentId: String?,
        path: String?,
        value: String?,
        context: Map<String, String>?
    ) {
        Log.d(TAG, "sendEvent: surfaceId=$surfaceId eventType=$eventType name=$name")
        try {
            val body = buildJsonObject {
                put("surface_id", surfaceId)
                put("event_type", eventType)
                name?.let { put("name", it) }
                sourceComponentId?.let { put("source_component_id", it) }
                path?.let { put("path", it) }
                value?.let { put("value", it) }
                context?.let { ctx ->
                    put("context", buildJsonObject { ctx.forEach { (k, v) -> put(k, v) } })
                }
            }

            val requestBody = body.toString().toRequestBody(jsonMediaType)
            val request = Request.Builder()
                .url("$baseUrl/event")
                .post(requestBody)
                .build()

            withContext(Dispatchers.IO) {
                client.newCall(request).execute().close()
            }
            Log.d(TAG, "sendEvent: success")
        } catch (e: Exception) {
            // Event sending is best-effort — don't crash the app
            Log.e(TAG, "sendEvent: failed", e)
        }
    }

    override fun sendFeedbackStream(
        messageId: String,
        rating: String,
        reason: String?
    ): Flow<StreamEvent> = flow {
        Log.i(TAG, "[feedback-stream] start: messageId=$messageId rating=$rating reason=$reason")
        try {
            val body = buildJsonObject {
                put("surface_id", messageId)
                put("event_type", "feedback")
                put("name", rating)
                reason?.let { put("value", it) }
            }
            val request = Request.Builder()
                .url("$baseUrl/event")
                .post(body.toString().toRequestBody(jsonMediaType))
                .header("Accept", "text/event-stream")
                .build()

            val response = withContext(Dispatchers.IO) {
                streamingClient.newCall(request).execute()
            }
            Log.d(TAG, "[feedback-stream] HTTP ${response.code}")

            if (!response.isSuccessful) {
                response.close()
                emit(StreamEvent.Error("HTTP ${response.code}: ${response.message}"))
                return@flow
            }

            val source = response.body?.source()
            if (source == null) {
                response.close()
                emit(StreamEvent.Error("Empty response body"))
                return@flow
            }

            try {
                var eventType = ""
                var doneReceived = false
                while (!doneReceived) {
                    val line = withContext(Dispatchers.IO) {
                        if (source.exhausted()) null else source.readUtf8Line()
                    } ?: break

                    when {
                        line.startsWith("event: ") -> {
                            eventType = line.removePrefix("event: ").trim()
                        }
                        line.startsWith("data: ") -> {
                            val data = line.removePrefix("data: ")
                            Log.d(TAG, "[feedback-stream] event=$eventType dataLen=${data.length}")
                            when (eventType) {
                                "a2ui_op" -> emit(StreamEvent.A2UiOp(data))
                                "text" -> {
                                    try {
                                        val textObj = json.parseToJsonElement(data).jsonObject
                                        val text = textObj["text"]?.jsonPrimitive?.contentOrNull ?: ""
                                        emit(StreamEvent.TextContent(text))
                                    } catch (e: Exception) {
                                        Log.w(TAG, "[feedback-stream] malformed text event", e)
                                    }
                                }
                                "done" -> {
                                    val doneMessage = parseDoneEvent(data)
                                    emit(StreamEvent.Done(doneMessage))
                                    doneReceived = true
                                }
                                "token" -> {
                                    try {
                                        val tokenData = json.decodeFromString<TokenData>(data)
                                        emit(StreamEvent.Token(tokenData.token))
                                    } catch (e: Exception) {
                                        Log.w(TAG, "[feedback-stream] malformed token event", e)
                                    }
                                }
                            }
                            eventType = ""
                        }
                    }
                }
            } finally {
                try { response.close() } catch (_: Exception) {}
            }
        } catch (e: Exception) {
            Log.e(TAG, "[feedback-stream] error", e)
            emit(StreamEvent.Error(e.message ?: "Feedback streaming error"))
        }
    }

    override fun getGreeting(): String {
        val hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY)
        return when {
            hour < 12 -> "morning"
            hour < 17 -> "afternoon"
            else -> "evening"
        }
    }
}

File: ios/A2UIChatApp/Data/Repository/RealChatRepository.swift
import Foundation

class RealChatRepository: ChatRepository {
    private let baseURL = "http://127.0.0.1:8000"

    private lazy var sseSession: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 600   // 10 min — LLM can be slow
        config.timeoutIntervalForResource = 600
        return URLSession(configuration: config)
    }()

    func sendMessageStream(message: String) -> AsyncThrowingStream<StreamEvent, Error> {
        print("[A2UI.Repo] sendMessageStream called, message=\(message.prefix(40))")
        return AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    print("[A2UI.Repo] inner Task starting performStream")
                    try await self.performStream(
                        url: URL(string: "\(self.baseURL)/chat/stream")!,
                        body: ["message": message],
                        continuation: continuation
                    )
                } catch {
                    print("[A2UI.Repo] performStream error: \(error)")
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in
                task.cancel()
            }
        }
    }

    private func performStream(
        url: URL,
        body: [String: Any],
        continuation: AsyncThrowingStream<StreamEvent, Error>.Continuation
    ) async throws {
        print("[A2UI.Repo] performStream: \(url)")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        print("[A2UI.Repo] opening bytes stream to \(url)")
        let (bytes, response) = try await sseSession.bytes(for: request)
        print("[A2UI.Repo] response received: \(response)")

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }
        print("[A2UI.Repo] HTTP 200 OK, starting line iteration")

        var currentEventType = ""
        var streamDone = false

        for try await line in bytes.lines {
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)

            if trimmed.hasPrefix("event: ") {
                currentEventType = String(trimmed.dropFirst("event: ".count))
            } else if trimmed.hasPrefix("data: ") {
                let dataStr = String(trimmed.dropFirst("data: ".count))

                switch currentEventType {
                case "a2ui_op":
                    print("[A2UI.Repo] a2ui_op: \(dataStr.prefix(80))")
                    continuation.yield(.a2uiOp(dataStr))
                case "text":
                    print("[A2UI.Repo] text event")
                    if let raw = dataStr.data(using: .utf8),
                       let json = try? JSONSerialization.jsonObject(with: raw) as? [String: Any],
                       let text = json["text"] as? String {
                        continuation.yield(.textContent(text))
                    }
                case "token":
                    if let raw = dataStr.data(using: .utf8),
                       let json = try? JSONSerialization.jsonObject(with: raw) as? [String: Any],
                       let token = json["token"] as? String {
                        continuation.yield(.token(token))
                    }
                case "done":
                    print("[A2UI.Repo] done")
                    continuation.yield(.done(nil))
                    streamDone = true
                    break
                default:
                    if !currentEventType.isEmpty {
                        print("[A2UI.Repo] unknown event: '\(currentEventType)'")
                    }
                }
                currentEventType = ""
                if streamDone { break }
            } else if trimmed.isEmpty {
                // SSE event boundary — reset
                currentEventType = ""
            }
            // Lines starting with ":" are SSE comments (e.g. ping) — ignore
        }

        if !streamDone {
            continuation.yield(.done(nil))
        }
        print("[A2UI.Repo] stream finished, streamDone=\(streamDone)")
        continuation.finish()
    }

    func sendEvent(
        surfaceId: String,
        eventType: String,
        name: String,
        sourceComponentId: String,
        path: String?,
        value: String?,
        context: [[String: String]]?
    ) async {
        guard let url = URL(string: "\(baseURL)/event") else { return }
        var body: [String: Any] = [
            "surface_id": surfaceId,
            "event_type": eventType,
            "name": name,
            "source_component_id": sourceComponentId
        ]
        if let path = path { body["path"] = path }
        if let value = value { body["value"] = value }
        if let context = context { body["context"] = context }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        _ = try? await URLSession.shared.data(for: request)
    }

    func sendFeedbackStream(messageId: String, rating: String, reason: String?) -> AsyncThrowingStream<StreamEvent, Error> {
        // Feedback uses the same /event endpoint which requires surface_id.
        // We send it as a simple fire-and-forget POST (no SSE response needed).
        var context: [String: String] = ["rating": rating]
        if let reason = reason { context["reason"] = reason }

        let body: [String: Any] = [
            "surface_id": "",          // no surface context for feedback
            "event_type": "feedback",
            "name": messageId,
            "context": context
        ]

        return AsyncThrowingStream { continuation in
            let task = Task {
                guard let url = URL(string: "\(self.baseURL)/event") else {
                    continuation.finish()
                    return
                }
                var request = URLRequest(url: url)
                request.httpMethod = "POST"
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                request.httpBody = try? JSONSerialization.data(withJSONObject: body)
                _ = try? await URLSession.shared.data(for: request)
                continuation.yield(.done(nil))
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    // MARK: — Spec-compliant JSONL streaming

    /// Set to `true` to hit `/chat/stream/jsonl` instead of `/chat/stream`.
    static let useJsonlEndpoint = false

    /// Spec-compliant JSONL stream from `/chat/stream/jsonl`.
    ///
    /// All SSE events arrive as plain `data:` lines (no custom `event:` type).
    /// Each line is a JSONL object dispatched on its top-level key:
    ///   - `"text"`            → StreamEvent.textContent
    ///   - `"surfaceUpdate"` / `"dataModelUpdate"` / `"beginRendering"` → StreamEvent.a2uiOp
    ///   - `"done"`            → StreamEvent.done
    func sendMessageStreamJsonl(message: String) -> AsyncThrowingStream<StreamEvent, Error> {
        print("[A2UI.Repo] sendMessageStreamJsonl called, message=\(message.prefix(40))")
        return AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    try await self.performJsonlStream(message: message, continuation: continuation)
                } catch {
                    print("[A2UI.Repo] performJsonlStream error: \(error)")
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    private func performJsonlStream(
        message: String,
        continuation: AsyncThrowingStream<StreamEvent, Error>.Continuation
    ) async throws {
        let url = URL(string: "\(baseURL)/chat/stream/jsonl")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["message": message])

        print("[A2UI.Repo] opening JSONL stream to \(url)")
        let (bytes, response) = try await sseSession.bytes(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }
        print("[A2UI.Repo] JSONL HTTP 200 OK")

        var streamDone = false
        for try await line in bytes.lines {
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            guard trimmed.hasPrefix("data: ") else { continue }

            let dataStr = String(trimmed.dropFirst("data: ".count))
            if dataStr.isEmpty || dataStr == "{}" { continue }

            guard let raw = dataStr.data(using: .utf8),
                  let json = try? JSONSerialization.jsonObject(with: raw) as? [String: Any] else {
                print("[A2UI.Repo] JSONL parse error on: \(dataStr.prefix(100))")
                continue
            }

            if let text = (json["text"] as? String) {
                print("[A2UI.Repo] JSONL text: \(text.prefix(60))")
                continuation.yield(.textContent(text))
            } else if json["surfaceUpdate"] != nil || json["dataModelUpdate"] != nil || json["beginRendering"] != nil {
                let key = json.keys.first ?? "?"
                print("[A2UI.Repo] JSONL a2ui op: \(key) len=\(dataStr.count)")
                continuation.yield(.a2uiOp(dataStr))
            } else if json["done"] != nil {
                print("[A2UI.Repo] JSONL done")
                continuation.yield(.done(nil))
                streamDone = true
                break
            } else {
                print("[A2UI.Repo] JSONL unknown key: \(json.keys)")
            }
        }

        if !streamDone {
            continuation.yield(.done(nil))
        }
        print("[A2UI.Repo] JSONL stream finished, streamDone=\(streamDone)")
        continuation.finish()
    }
}
